repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,360 @@
+
+                  val impersonatorId = validToken.impersonatorId.map(iid => OpenStackServiceHeader.IMPERSONATOR_ID.toString -> iid)
+                  val impersonatorName = validToken.impersonatorName.map(in => OpenStackServiceHeader.IMPERSONATOR_NAME.toString -> in)
+
+                  val identityStatus = OpenStackServiceHeader.IDENTITY_STATUS.toString -> IdentityStatus.Confirmed.toString
+
+                  Pass(headers ++ userHeaders ++ rolesHeader + tenantName + xAuthHeader ++ defaultRegion ++ contactId
+                    + expirationDate ++ impersonatorId ++ impersonatorName + identityStatus)
+                case reject: Reject => reject
+              }
+
+              addHeaders
+            case Failure(x: IdentityAdminTokenException) =>
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+            case Failure(x: IdentityCommuncationException) =>
+              Reject(SC_BAD_GATEWAY, failure = Some(x))
+            case Failure(x) =>
+              //TODO: this isn't yet complete, should return other status codes?
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+          }
+        } getOrElse {
+          //Don't have an auth token to validate
+          Reject(SC_FORBIDDEN, Some("Auth token not found in headers"))
+        }
+      }
+
+      //Handle the result of the filter
+      result match {
+        case rejection: Reject =>
+          val message: Option[String] = rejection match {
+            case Reject(_, Some(x), _) => Some(x)
+            case Reject(code, None, Some(failure)) => Some(failure.getMessage)
+            case _ => None
+          }
+          //Handle delegation if necessary
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status ${rejection.status}")
+              val delegationHeaders = buildDelegationHeaders(rejection.status,
+                "keystone-v2",
+                message.getOrElse("Failure in the Keystone v2 filter").replace("\n", ""),
+                delegating.getQuality)
+
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+              request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS.toString, IdentityStatus.Indeterminate.toString)
+
+              chain.doFilter(request, response)
+            case None =>
+              logger.debug(s"Rejecting with status ${rejection.status}")
+              response.setHeader(CommonHttpHeader.WWW_AUTHENTICATE.toString, s"Keystone uri=${config.getIdentityService.getUri}")
+              message match {
+                case Some(m) =>
+                  logger.debug(s"Rejection message: $m")
+                  response.sendError(rejection.status, m)
+                case None => response.sendError(rejection.status)
```

Comment:
Should we use sendError here? sendError commits the response stream meaning previous filters in the filter chain cannot modify that portion of the response. I know we use sendError in a couple of places, so I thought I'd call attention to this.

Of course, if a previous filter uses our new response wrapper, this is a non-issue. It's only the case where a filter is mutating the response directly that we /might/ run into trouble. Now to a standard filter writer, that may be the approach they take since a response is mutable by design.